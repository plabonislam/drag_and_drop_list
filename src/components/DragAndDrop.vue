<script setup>
import { ref } from 'vue'
import HeaderList from './HeaderList.vue'

const dragItem = ref(null)
const dragOverItem = ref(null)

const items = ref([
  { fullname: 'first', age: 20 },
  { fullname: 'second', age: 20 },
  { fullname: 'third', age: 20 },
  {
    fullname: 'fourth',
    age: 20
  }
])

const props = defineProps({
  title: {
    type: String,
    default: 'Drag and Drop List'
  }
})

const onDragStart = (e) => {
  dragItem.value = e.target.id
}

const onDragEnter = (e) => {
  dragOverItem.value = e.currentTarget.id
}

const onDragend = (e) => {
  const copyItems = [...items.value]
  const dragItemContent = copyItems[dragItem.value]
  copyItems.splice(dragItem.value, 1)
  copyItems.splice(dragOverItem.value, 0, dragItemContent)
  items.value = [...copyItems]
}

</script>
<template>
  <p class="title">{{ props.title }}</p>
  <button disabled="true" class="bar" draggable="true">btn</button>
  <table style="width: 100%">
    <HeaderList />
    <tr
      v-for="(item, index) in items"
      :key="index"
      :id="index"
      draggable="true"
      disabled="true"
      @dragstart="onDragStart"
      @dragenter="onDragEnter"
      @dragend="onDragend"
    >
      <td>{{ item.fullname }}</td>
      <td>{{ item.age }}</td>
    </tr>
  </table>
</template>

<style scoped>
th,
td {
  border: 1px solid black;
  text-align: center;
  padding: 10px 10px;
  width: 400px;
}

td:draggable {
  opacity: 0.5;
  background: red;
}
</style>
