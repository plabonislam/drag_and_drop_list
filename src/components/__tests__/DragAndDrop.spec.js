import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import DragAndDrop from '../DragAndDrop.vue'

describe('DragAndDrop', () => {
    it('renders properly', () => {
        const wrapper = mount(DragAndDrop)
        expect(wrapper.html()).toMatchSnapshot();
    })

    it('title checking with props', () => {
        const wrapper = mount(DragAndDrop, {
            props: {
                title: 'Drag and Drop List'
            }
        })
        const titleElement = wrapper.find('.title') // Adjust the selector to match your title element
        expect(titleElement.text()).toBe('Drag and Drop List');
        })

        it('title checking without props', () => {
        const wrapper = mount(DragAndDrop)
        expect(wrapper.html()).toContain('Drag and Drop List'); // New way to check the title
        })

    it('draggable button checking', () => {
        const wrapper = mount(DragAndDrop)
        const button = wrapper.find('button')
        expect(button.exists()).toBe(true);
        expect(button.attributes('draggable')).toBe('true');
    })

    it('renders list items', () => {
        const wrapper = mount(DragAndDrop)
        const listItems = wrapper.findAll('tr')
        expect(listItems.length).toBeGreaterThan(0);
    })

    // it('list items are draggable', () => {
    //     const wrapper = mount(DragAndDrop)
    //     const listItems = wrapper.findAll('tr')
    //     listItems.forEach(item => {
    //         expect(item.attributes('draggable')).toBe('true');
    //     })
    // })

    // it('emits event on drag start', async () => {
    //     const wrapper = mount(DragAndDrop)
    //     const listItem = wrapper.find('li')
    //     await listItem.trigger('dragstart')
    //     expect(wrapper.emitted()).toHaveProperty('dragstart')
    // })

    // it('emits event on drop', async () => {
    //     const wrapper = mount(DragAndDrop)
    //     const listItem = wrapper.find('li')
    //     await listItem.trigger('drop')
    //     expect(wrapper.emitted()).toHaveProperty('drop')
    // })
})
