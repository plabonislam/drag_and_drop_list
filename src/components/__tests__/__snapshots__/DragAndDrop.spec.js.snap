// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DragAndDrop > renders properly 1`] = `
"<p data-v-6722d57b="" class="title">Drag and Drop List</p>
<button data-v-6722d57b="" disabled="" class="bar" draggable="true">btn</button>
<table data-v-6722d57b="" style="width: 100%;">
  <div><input type="text" name="t1" class="foo"><input type="text" name="t2" class="foo"></div>
  <tr>
    <th>TO DO</th>
    <th>IN PROGRESS</th>
    <th>QA READY</th>
    <th>QA</th>
    <th>RELEASE READY</th>
    <th>DONE</th>
  </tr>
  <tr data-v-6722d57b="" id="0" draggable="true" disabled="true">
    <td data-v-6722d57b="">first</td>
    <td data-v-6722d57b="">20</td>
  </tr>
  <tr data-v-6722d57b="" id="1" draggable="true" disabled="true">
    <td data-v-6722d57b="">second</td>
    <td data-v-6722d57b="">20</td>
  </tr>
  <tr data-v-6722d57b="" id="2" draggable="true" disabled="true">
    <td data-v-6722d57b="">third</td>
    <td data-v-6722d57b="">20</td>
  </tr>
  <tr data-v-6722d57b="" id="3" draggable="true" disabled="true">
    <td data-v-6722d57b="">fourth</td>
    <td data-v-6722d57b="">20</td>
  </tr>
</table>"
`;
